src/fileformats/ocd_georef_fields.cpp

```diff
@@ -0,0 +1,457 @@
+	// parameter - it can be an instance of UTMZoneParameter, IntRangeParameter
+	// (with "factor" and "bias") or FullSpecParameter
+	auto param = crs_template->parameters()[0];
+
+	// run every parameter value through the parameter processor which
+	// converts it into a string list that can be consumed by the above
+	// CRS specification template
+	const QString crs_param_string =
+	        result.crs_id != MapperCrs::Utm || result.crs_param >= 0
+	        ? QString::number(result.crs_param) // common case, plain number
+	        : QString::number(-result.crs_param) + QLatin1String(" S"); // UTM South zones
+
+	for (const auto& spec_value : param->specValues(crs_param_string))
+		spec = spec.arg(spec_value);
+
+	// `spec` and the pair <`crs_id_string`, `crs_param_string`> both describe
+	// the same thing at this place
+	georef.setProjectedCRS(crs_id_string, spec, { crs_param_string });
+}
+
+/**
+ * OcdCrsRecord struct is used for passing CRS information around within the
+ * translation code.
+ */
+struct OcdCrsRecord
+{
+	const OcdGrid ocd_grid_id; ///< OCD grid number
+	const int ocd_zone_id; ///< OCD zone number, may be negative
+};
+
+/**
+ * Provide an OCD grid/zone combinations for a Mapper georeferencing.
+ *
+ * @see translateOcdToMapper
+ */
+OcdCrsRecord translateMapperToOcd(const MapperCrs crs_unique_id,
+                                  const int crs_param)
+{
+	switch (crs_unique_id)
+	{
+	case MapperCrs::Local:
+		return { OcdGrid::Local, 0 };
+		break;
+
+	case MapperCrs::Utm:
+		if (qAbs(crs_param) >= 1 && qAbs(crs_param) <= 60)
+			return { OcdGrid::Utm, crs_param };
+		break;
+
+	case MapperCrs::GaussKrueger:
+		if (crs_param >= 2 && crs_param <= 5)
+			return { OcdGrid::Germany, crs_param };
+		break;
+
+	case MapperCrs::Epsg:
+		// handle UTM separately, the rest is in the table
+		if (crs_param >= 32601 && crs_param <= 32660)
+			return { OcdGrid::Utm, crs_param - 32600 };
+		else if (crs_param >= 32701 && crs_param <= 32760)
+			return { OcdGrid::Utm, 32700 - crs_param };
+
+		for (const auto& record : grid_zone_epsg_codes)
+		{
+			if (record.epsg_code == crs_param)
+				return { record.ocd_grid_id, record.ocd_zone_id };
+		}
+		break;
+	default:
+		qWarning("Internal program error: translateMapperToOcd() should not"
+		         " be fed with invalid values (%d)", static_cast<int>(crs_unique_id));
+	}
+
+	return { OcdGrid::Invalid, 0 };
+}
+
+}  // anonymous namespace
+
+bool operator==(const OcdGeorefFields& lhs, const OcdGeorefFields& rhs)
+{
+	return lhs.i == rhs.i
+	        && lhs.m == rhs.m
+	        && lhs.x == rhs.x
+	        && lhs.y == rhs.y
+	        && ((qIsNaN(lhs.a) && qIsNaN(rhs.a)) || qAbs(lhs.a - rhs.a) < 5e-9) // 8-digit precision or both NaN's
+	        && lhs.r == rhs.r;
+}
+
+void OcdGeorefFields::setupGeoref(Georeferencing& georef,
+                                  const std::function<void (const QString&)>& warning_handler) const
+{
+	if (m >= 0)
```

Comment:
`m >= 1`

Reply:
Fun to watch you fix your own code. This line came from commit 630987577 through a few reviews and rewrites. Only in "OCD: Make OcdGeoref even more isolated" `scale` was renamed to `m`. Will apply the change.

Reply:
It would be much less funny if I wouldn't notice the errors I introduced... Feel free to fix them faster than I can ;-)